compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -311,6 +314,11 @@ public void visit(BLangLiteral literalExpr) {
                 return;
             }
         }
+
+        if (literalType.tag == TypeTags.NIL && "null".equals(literalExpr.originalValue)) {
```

Comment:
As we discussed offline we have to handle scenarios with finite type, where nil is applicable.

Reply:
Addressed with 836bae5